src/Avalonia.Base/Visual.cs

```diff
@@ -495,10 +515,13 @@ protected virtual void OnAttachedToVisualTreeCore(VisualTreeAttachmentEventArgs
             OnAttachedToVisualTree(e);
             AttachedToVisualTree?.Invoke(this, e);
             InvalidateVisual();
+            
             _visualRoot.Renderer.RecalculateChildren(_visualParent!);
-
-            if (ZIndex != 0 && VisualParent is Visual parent)
-                parent.HasNonUniformZIndexChildren = true;
+            
+            if (ZIndex != 0 && _visualParent is { })
+                _visualParent.HasNonUniformZIndexChildren = true;
+            
+            SyncIsEffectivelyVisible(_visualParent);
```

Comment:
Question: should `IsEffectivelyVisible` be set when the `AttachedToVisualTree`/`DetachedFromVisualTree` event is called? My initial thought is "yes", though you may know more than me.